File: Backend/src/main/scala/de/tkip/sbpm/persistence/query/GroupsRoles.scala
/*
 * S-BPM Groupware v1.2
 *
 * http://www.tk.informatik.tu-darmstadt.de/
 *
 * Copyright 2013 Telecooperation Group @ TU Darmstadt
 * Contact: Stephan.Borgert@cs.tu-darmstadt.de
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package de.tkip.sbpm.persistence.query

import de.tkip.sbpm.model.GroupRole

/**
 * PersistenceActor queries for "GroupsRoles".
 */
object GroupsRoles {
  // used to identify all GroupsRoles queries
  trait Query extends BaseQuery

  object Read {
    def apply() = All
    /**
     * returns all group -> role associations (Seq[GroupRole])
     */
    case object All extends Query
    /**
     * returns group -> role association by group and role id
     * or None if not found (Option[GroupRole])
     */
    case class ById(groupId: Int, roleId: Int) extends Query
    /**
     * returns all group -> role associations for a group id
     * (Seq[GroupRole])
     */
    case class ByGroupId(id: Int) extends Query
    /**
     * returns all group -> role associations for a role id
     * (Seq[GroupRole])
     */
    case class ByRoleId(id: Int) extends Query
  }

  object Save {
    def apply(groupRole: GroupRole*) = Entity(groupRole: _*)
    /**
     * saves all given associations
     * if one entity given, returns (groupId, roleId) if entry was created
     * or None if entry was updated (Option[(Int, Int)])
     * if multiple entities given, Seq[Option[(Int, Int)]]
     * is returned respectively
     */
    case class Entity(groupRole: GroupRole*) extends Query
  }

  object Delete {
    def apply(groupRole: GroupRole) = ById(groupRole.groupId, groupRole.roleId)
    /**
     * deletes association by group and role id with empty result
     */
    case class ById(groupId: Int, roleId: Int) extends Query
    /**
     * deletes associations of group with empty result
     */
    case class ByGroupId(id: Int) extends Query
    /**
     * deletes associations of role with empty result
     */
    case class ByRoleId(id: Int) extends Query
  }
}

File: ProcessManagement/js/models/groupsRoles.js
/*
 * S-BPM Groupware v1.2
 *
 * http://www.tk.informatik.tu-darmstadt.de/
 *
 * Copyright 2013 Telecooperation Group @ TU Darmstadt
 * Contact: Stephan.Borgert@cs.tu-darmstadt.de
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/.
 */
define([
	"knockout",
	"model",
	"underscore"
], function( ko, Model, _ ) {

	// Our main model that will be returned at the end of the function.
	//
	// Process is responsivle for everything associated with processes directly.
	//
	// For example: Getting a list of all processes, savin a process,
	// validating the current process etc.
	GroupsRoles = Model( "GroupsRoles", { methods: { create: "PUT" } } );

	GroupsRoles.attrs({
		isActive: {
			type: "boolean",
			defaults: true
		},
		roleId: "integer",
		groupId: "integer"
	});

	GroupsRoles.ids([ "roleId", "groupId" ]);

	GroupsRoles.belongsTo( "group" );
	GroupsRoles.belongsTo( "role" );

	// Group.extend({);

	GroupsRoles.include({
		// Initialize is a special method defined as an instance method.  If any
		// method named "initializer" is given, it will be called upon object
		// creation (when calling new model()) with the context of the model.
		// That is, "this" refers to the model itself.
		// This makes it possible to define defaults for attributes etc.
		initialize: function( data ) {
			if ( !data ) {
				data = {};
			}

			// Set some defaults for the data object (used as a hash)
			_( data ).defaults({
				isActive: true
			});

			this.isActive( data.isActive );
		}

		// Custom validator object. Validators are (like the initialize function)
		// special in a sense that this object will be iterated over when the
		// "validate" method is executed.
		// validators: { }
	});

	return GroupsRoles;
});
